File: packages/cli-python/mcpseal/diff.py
# Mirrors packages/cli-node/src/diff.ts. build-bible.md Part 3.2, Tasks.md
# 2.6: `mcpseal diff` — human-readable old-vs-new for any drifted tool.
# Only descriptions can be text-diffed (the lockfile stores description,
# not the full tool object — Tasks.md 2.3 Change Log). When the hash
# differs but the description text is identical, the change must be in
# inputSchema — say so honestly rather than fabricate a schema diff.
from __future__ import annotations

from typing import TypedDict

from mcpseal.scan import ScanDecision, scan


class DriftDiff(TypedDict):
    serverName: str
    toolName: str
    oldDescription: str
    newDescription: str
    descriptionChanged: bool
    possibleSchemaChange: bool


def diff_drifted(project_dir: str, lockfile_path: str | None = None) -> list[DriftDiff]:
    decisions: list[ScanDecision] = scan(project_dir, lockfile_path)
    diffs: list[DriftDiff] = []
    for d in decisions:
        if d["result"].get("reason") != "blocked_drift":
            continue
        old_description = d["result"].get("oldDescription", "")
        new_description = d["result"].get("newDescription", "")
        diffs.append(
            {
                "serverName": d["serverName"],
                "toolName": d["toolName"],
                "oldDescription": old_description,
                "newDescription": new_description,
                "descriptionChanged": old_description != new_description,
                "possibleSchemaChange": old_description == new_description,
            }
        )
    return diffs


def format_diff(diff: DriftDiff) -> str:
    lines = [f"{diff['serverName']}/{diff['toolName']}:"]
    if diff["descriptionChanged"]:
        lines.append(f"  - {diff['oldDescription']}")
        lines.append(f"  + {diff['newDescription']}")
    if diff["possibleSchemaChange"]:
        lines.append("  (description unchanged — the change is in inputSchema; full schema diff not available)")
    return "\n".join(lines)

File: packages/cli-node/src/diff.ts
// build-bible.md Part 3.2, docs/Tasks.md 2.6: `mcpseal diff` — human-readable
// old-vs-new for any drifted tool. Can only text-diff descriptions, since
// that's the only pre-mutation state the lockfile stores (docs/Tasks.md 2.3
// Change Log deliberately chose description-only over storing the full
// {name,description,inputSchema} object). When the hash differs but the
// description text is identical, the change must be in inputSchema — flag
// that honestly rather than fabricate a schema diff we have no data for.
import { scan, type ScanDecision } from "./scan.js";

export interface DriftDiff {
  serverName: string;
  toolName: string;
  oldDescription: string;
  newDescription: string;
  descriptionChanged: boolean;
  possibleSchemaChange: boolean;
}

export async function diffDrifted(projectDir: string, lockfilePath?: string): Promise<DriftDiff[]> {
  const decisions: ScanDecision[] = await scan(projectDir, lockfilePath);
  return decisions
    .filter((d) => d.result.reason === "blocked_drift")
    .map((d) => {
      const oldDescription = d.result.oldDescription ?? "";
      const newDescription = d.result.newDescription ?? "";
      return {
        serverName: d.serverName,
        toolName: d.toolName,
        oldDescription,
        newDescription,
        descriptionChanged: oldDescription !== newDescription,
        possibleSchemaChange: oldDescription === newDescription,
      };
    });
}

export function formatDiff(diff: DriftDiff): string {
  const lines = [`${diff.serverName}/${diff.toolName}:`];
  if (diff.descriptionChanged) {
    lines.push(`  - ${diff.oldDescription}`);
    lines.push(`  + ${diff.newDescription}`);
  }
  if (diff.possibleSchemaChange) {
    lines.push(`  (description unchanged — the change is in inputSchema; full schema diff not available)`);
  }
  return lines.join("\n");
}
